### PR TITLE
feat: create blockFoundCh in RetrieveLoop to enable syncing faster than the DA block time

### LINKED ISSUE #1193: Figure out how to modify RetrieveLoop to catch up with latest DA height faster
Currently, the `RetrieveLoop` in Block Manager is responsible for retrieving blocks from the DA layer. It maintains a local version of the DA height, retrieves blocks from that height once every DA block time (~15 seconds) and increments that height. This retrieve mechanism is used for hard confirmations of blocks.

We'd like `RetrieveLoop` to keep up with the latest DA height faster to eliminate this lag.  One way to do this would be to get the latest height, and then retrieve from all heights until that latest height starting from the last height checked plus one. Something similar happens in `BlockRetrieveLoop` currently. Although, we do not have a way to query the DA layer for the latest height in the DA interface so it's not straightforward implement this. 

The `submitBlocksToDA` method does have access to the height that the blocks were published on which is more in sync with the DA layer than the local height so we could use that but this would only keep the height updated for a block producer and not validating full nodes which do not submit blocks to the DA layer. We should figure out an alternative way to pull blocks from all DA heights until the latest height for all types of nodes.

### PR DESCRIPTION
**Pull request by MSevey**
## Overview

closes #1193 

cc @Manav-Aggarwal  @tuxcanfly 

## Summary by CodeRabbit

Given the extensive details provided, I've distilled the changes into a more user-friendly format:

- **New Features**
  - Added functionality to check if a block is hard confirmed.
  - Implemented a channel to track successful block retrieval.

- **Bug Fixes**
  - Correctly incremented the counter after block retrieval.

- **Tests**
  - Added a test to ensure nodes can sync Data Availability blocks faster than the block time.
  - Introduced a test to verify the behavior of the `GetHardConfirmation` method.

- **Refactor**
  - Removed an empty line in the `OnStart` function.
  - Adjusted variable and function initialization for testing purposes.

- **Chores**
  - Implemented a function to safely close a channel if not already closed.

This summary focuses on the user-facing impact of the changes, omitting internal details.

### CODE REVIEW

**Comment by gupadhyaya** [comment]
@MSevey i feel like this solution only partially helps. is there a way to subscribe (possibly via da-node) to DA headers and then trigger? May be `Subscribe` in https://node-rpc-docs.celestia.org/?version=v0.11.0#header

**Comment by MSevey** [comment]
> @MSevey i feel like this solution only partially helps. is there a way to subscribe (possibly via da-node) to DA headers and then trigger? May be `Subscribe` in https://node-rpc-docs.celestia.org/?version=v0.11.0#header

@gupadhyaya 

I think that would require a larger refactor right? And only really help in the lag of the most recent block. 

If I'm understanding the rollkit sync currently, we only sync forward, i.e. sync block 1 then 2 then 3, etc. 

The subscribe method in the node would help us get the latest height, but if the rollkit node is at height 10 and the DA is at height 20, would the single trigger of the retrieve loop pull all the rollup blocks from 10 to 20? Or would we need to still iteratively sync 10,11,12,...,20?

I think it is a reasonable optimization to do, but should we align that with the DA interface?
We should probably avoid direct node API calls outside the DA interface right?

This could be a good follow up for coordinating with the DA interface.

**Comment by Manav-Aggarwal** [comment]
> > @MSevey i feel like this solution only partially helps. is there a way to subscribe (possibly via da-node) to DA headers and then trigger? May be `Subscribe` in https://node-rpc-docs.celestia.org/?version=v0.11.0#header
> 
> @gupadhyaya
> 
> I think that would require a larger refactor right? And only really help in the lag of the most recent block.
> 
> If I'm understanding the rollkit sync currently, we only sync forward, i.e. sync block 1 then 2 then 3, etc.
> 
> The subscribe method in the node would help us get the latest height, but if the rollkit node is at height 10 and the DA is at height 20, would the single trigger of the retrieve loop pull all the rollup blocks from 10 to 20? Or would we need to still iteratively sync 10,11,12,...,20?
> 
> I think it is a reasonable optimization to do, but should we align that with the DA interface? We should probably avoid direct node API calls outside the DA interface right?
> 
> This could be a good follow up for coordinating with the DA interface.

Agreed, this should be part of a larger refactor that requires changes on the DA interface as well. The current design does not support `Subscribe` as the mechanism to sync from DA especially when a new Rollkit full node is catching up to the latest rollup height.

File: block/block_cache_test.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	_, ok = bc.getBlock(height)
+	require.False(ok, "getBlock should return false after deleteBlock")
+
+	// Test isSeen and setSeen
+	require.False(bc.isSeen("hash"), "isSeen should return false for unseen hash")
+	bc.setSeen("hash")
+	require.True(bc.isSeen("hash"), "isSeen should return true for seen hash")
+
+	// Test setHardConfirmed
+	require.False(bc.hardConfirmations["hash"], "hardConfirmations should be false for unseen hash")
```
Comment (tuxcanfly) [inline block/block_cache_test.go:69]:
```suggestion
	require.False(bc.isHardConfirmed("hash"), "hardConfirmations should be false for unseen hash")
```

File: block/block_cache_test.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	bc.setHardConfirmed("hash")
+	require.True(bc.hardConfirmations["hash"], "hardConfirmations should be true for seen hash")
```
Comment (tuxcanfly) [inline block/block_cache_test.go:71]:
```suggestion
	require.True(bc.isHardConfirmed("hash"), "hardConfirmations should be true for seen hash")
```

File: block/block_cache_test.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/rollkit/rollkit/types"
+)
+
+// TODO: refactor this helper as it is defined in multiple places
+func getRandomBlock(height uint64, nTxs int) *types.Block {
```
Comment (Manav-Aggarwal) [inline block/block_cache_test.go:12]: Blocking this on https://github.com/rollkit/rollkit/pull/1306

**Review by Manav-Aggarwal — CHANGES_REQUESTED** [review]

File: block/manager.go [hunk 4]
```diff
@@ -459,18 +464,28 @@ func (m *Manager) getBlocksFromBlockStore(ctx context.Context, startHeight, endH
 
 // RetrieveLoop is responsible for interacting with DA layer.
 func (m *Manager) RetrieveLoop(ctx context.Context) {
+	// blockFoundCh is used to track when we successfully found a block so
+	// that we can continue to try and find blocks are the next DA height.
```
Comment (tuxcanfly) [inline block/manager.go:468]:
```suggestion
	// that we can continue to try and find blocks that are in the next DA height.
```

**Review by tuxcanfly — APPROVED** [review]
Agree about future `Subscribe` support. Found some nits.

**Comment by gupadhyaya** [comment]
@Manav-Aggarwal did you get a chance to test this on the gm app where you had observed the lag previously?

**Comment by Manav-Aggarwal** [comment]
> @Manav-Aggarwal did you get a chance to test this on the gm app where you had observed the lag previously?

Confirmed, tested gm app on main where it lags and then this branch where it syncs faster and catches up right after submitting those DA blocks on local-celestia-devnet

**Review by Manav-Aggarwal — APPROVED** [review]

File: block/manager.go [hunk 5]
```diff
@@ -459,18 +464,29 @@ func (m *Manager) getBlocksFromBlockStore(ctx context.Context, startHeight, endH
 
 // RetrieveLoop is responsible for interacting with DA layer.
 func (m *Manager) RetrieveLoop(ctx context.Context) {
+	// blockFoundCh is used to track when we successfully found a block so
+	// that we can continue to try and find blocks that are in the next DA height.
+	// This enables syncing faster than the DA block time.
+	blockFoundCh := make(chan struct{}, 1)
+	defer close(blockFoundCh)
```
Comment (Manav-Aggarwal) [inline block/manager.go:471]: Really elegant solution :)

**Review by Manav-Aggarwal — APPROVED** [review]
LGTM

File: node/full_client_test.go [hunk 6]
```diff
@@ -727,22 +728,23 @@ func createGenesisValidators(t *testing.T, numNodes int, appCreator func(require
 	dalc := &mockda.DataAvailabilityLayerClient{}
 	ds, err := store.NewDefaultInMemoryKVStore()
 	require.Nil(err)
-	err = dalc.Init([8]byte{}, nil, ds, log.TestingLogger())
+	err = dalc.Init([8]byte{}, nil, ds, test.NewFileLogger(t))
 	require.Nil(err)
 	err = dalc.Start()
 	require.Nil(err)
-	t.Cleanup(func() {
+	defer func() {
```
[the remark this answers is not in the document]
Comment (Manav-Aggarwal) [reply node/full_client_test.go]: Reverted.

**Review by Manav-Aggarwal — DISMISSED** [review]

File: node/full_node_integration_test.go [hunk 7]
```diff
@@ -189,6 +189,96 @@ func TestLazyAggregator(t *testing.T) {
...
+	defer func() {
+		require.NoError(node2.Stop())
+	}()
+
+	// Start and launch the timer in a go routine to ensure that the test
+	// fails if the nodes do not sync before the timer expires
+	ch := make(chan struct{})
+	defer safeClose(ch)
+	// Set the timer to expire 1 block before the numberOfBlocksToSyncTill
+	timer := time.NewTimer((numberOfBlocksToSyncTill - 1) * bmConfig.DABlockTime)
```
Comment (tuxcanfly) [inline node/full_node_integration_test.go:238]: We can probably reduce this to 2 blocks.

**Review by tuxcanfly — CHANGES_REQUESTED** [review]
Minor nits.

File: node/full_node_integration_test.go (hunk 7 above)
Comment (Manav-Aggarwal) [reply node/full_node_integration_test.go]: Reduced it to a little over 1 block to be stricter.

**Review by Manav-Aggarwal — APPROVED** [review]

**Review by tuxcanfly — APPROVED** [review]

**Pull request merged**